packages/flutter/lib/src/widgets/shortcuts.dart

```diff
@@ -284,68 +288,94 @@ class ShortcutManager extends ChangeNotifier with Diagnosticable {
-          pseudoKeys.add(synonyms.first);
-        } else {
-          pseudoKeys.add(setKey);
+      for (final KeyboardKey setKey in keysPressed.keys) {
+        if (setKey is LogicalKeyboardKey) {
+          final Set<LogicalKeyboardKey> synonyms = setKey.synonyms;
+          if (synonyms.isNotEmpty) {
+            // There currently aren't any synonyms that match more than one key.
+            assert(synonyms.length == 1, 'Unexpectedly encountered a key synonym with more than one key.');
+            pseudoKeys.add(synonyms.first);
+          } else {
+            pseudoKeys.add(setKey);
+          }
         }
       }
       matchedIntent = _shortcuts[LogicalKeySet.fromSet(pseudoKeys)];
     }
+    return matchedIntent;
+  }
+
+  /// Handles a key press `event` in the given `context`.
+  ///
+  /// The optional `keysPressed` argument is used as the set of currently
+  /// pressed keys. Defaults to a set derived from [RawKeyboard.keysPressed] if
+  /// `keysPressed` is not supplied.
+  ///
+  /// If a key mapping is found, then the associated action will be invoked
+  /// using the [Intent] that the `keysPressed` maps to, and the currently
+  /// focused widget's context (from [FocusManager.primaryFocus]).
+  ///
+  /// Returns a [KeyEventResult.handled] if an action was invoked, otherwise a
+  /// [KeyEventResult.stopPropagation] if [modal] is true, or if it maps to a
+  /// [DoNothingAction] with [DoNothingAction.shouldHandleKey] set to false, and
+  /// in all other cases returns [KeyEventResult.ignored].
+  ///
+  /// In order for an action to be invoked (and [KeyEventResult.handled]
+  /// returned), a pressed [KeySet] must be mapped to an [Intent], the [Intent]
+  /// must be mapped to an [Action], and the [Action] must be enabled.
+  @protected
+  KeyEventResult handleKeypress(
+    BuildContext context,
+    RawKeyEvent event, {
+    LogicalKeySet? keysPressed,
+  }) {
+    if (event is! RawKeyDownEvent) {
+      return KeyEventResult.ignored;
+    }
+    assert(context != null);
+    assert(keysPressed != null || RawKeyboard.instance.keysPressed.isNotEmpty,
+      'Received a key down event when no keys are in keysPressed. '
+      "This state can occur if the key event being sent doesn't properly "
+      'set its modifier flags. This was the event: $event and its data: '
+      '${event.data}');
+    final Intent? matchedIntent = find(keysPressed: keysPressed);
     if (matchedIntent != null) {
-      final BuildContext? primaryContext = primaryFocus?.context;
+      final BuildContext primaryContext = primaryFocus!.context!;
       assert (primaryContext != null);
-      Actions.invoke(primaryContext!, matchedIntent, nullOk: true);
-      return true;
+      final Action<Intent>? action = Actions.find<Intent>(
+        primaryContext,
+        intent: matchedIntent,
+        nullOk: true,
+      );
+      if (action != null && action.isEnabled(matchedIntent)) {
+        if (action is DoNothingAction) {
```

Comment:
I think in general we try to avoid hardcoding things like this. How would somebody go about implementing their own DoNothingAction?

Reply:
They would derive from `DoNothingAction`. I agree about hard coding the type, but other than adding an accessor to all actions to ask if they do nothing, I don't know how I would implement this.

Reply:
The Actions superclass could just have a default implementation for that accessor and only DoNothingAction overrides it. I think that would be preferred. 

I guess alternatively, invokeAction could also return something, but that may be too breaking?

Reply:
Well, `invokeAction` already returns the result of the action as an Object.  It would have to return some kind of container class instead, and I think that is probably too breaking.

The method would have to be on the `Action` class (`Actions` is the widget). OK, I'll see what that looks like.